File: Baekjoon/Graph/ABCDE/ABCDE.py
import sys
sys.stdin = open("Baekjoon/Graph/ABCDE/input.txt","r")
n,m = map(int,input().split(" "))
Graph = [[]for _ in range(n)]
for _ in range(m):
    a,b = map(int,input().split(" "))
    Graph[a].append(b)
    Graph[b].append(a)
    
visited = [False for _ in range(n)]

def DFS(now,cnt):
    visited[now] = True
    if cnt >= 4:
        print(1)
        exit()
    for next in Graph[now]:
        if not visited[next]:
            DFS(next,cnt+1)
            visited[next] = False

for i in range(n):
    DFS(i,0)
    visited[i] = False
print(0)
exit()

File: Baekjoon/Java/ABCDE.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.ArrayList;
import java.util.List;

public class ABCDE {  
    static List<Integer>[] list;
    static boolean[] visited;
    static boolean status;
    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("Baekjoon/java/input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        list = new ArrayList[n];
        for(int i =0;i<n;i++){
            list[i] = new ArrayList<>();
        }
        for(int i =0;i<m;i++){
            st = new StringTokenizer((br.readLine()));
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            list[a].add(b);
            list[b].add(a);
        }

        status = false;
        for(int i =0;i<n;i++){
            visited = new boolean[n];
            dfs(i,1);
            if(status){
                System.out.println(1);
                return;
            }
        }
        System.out.println(0);
    }
    static void dfs(int idx, int cnt){
        if(cnt == 5){
            status = true;
            return;
        }
        visited[idx] = true;
        for(int next : list[idx]){
            if(!visited[next]){
                dfs(next,cnt+1);
            }
        }
        visited[idx]=false;
    }
}